test/Linux/GstPlayerRefactorTest/src/GstPlayerTestApp.cpp

```diff
@@ -0,0 +1,316 @@
+    fs::path movieToLoad = mReloadPaths[ currentMovieIndex % 2 ];
+    currentMovieIndex++;
+    loadMovieFile( movieToLoad );
+}
+
+void GstPlayerTestApp::testStepForward()
+{
+    auto now = mTriggerTimer.getSeconds();
+    if( now - mTimeLastStepFwdTrigger >= mTriggerStepFwdRate ) {
+        newStepFwd();
+        mTimeLastStepFwdTrigger = now;
+        if( mRandomizeStepFwdRate ) mTriggerStepFwdRate = randFloat( .1f, .4f );
+    }
+}
+
+void GstPlayerTestApp::newStepFwd()
+{
+    if( ! mMovie ) return;
+    CI_LOG_I( "---------- NEW STEP FWD START ----------" );
+    CI_LOG_I( "POS BEFORE STEP : " << mMovie->getCurrentTime() );
+
+    // Pause the pipeline in order to get a better feeling of 
+    // the frame-by-frame action.
+    if( mMovie->isPlaying() ) mMovie->stop();
+
+    if( ! mMovie->stepForward() ) {
+        CI_LOG_W( "Failed to step forward!!!" );
+    }
+    // We have reached the end, so wrap and start over..
+    if( mMovie->getCurrentTime() == mLastStepPosition ) {
+        mMovie->seekToTime( 0.0f );
+    }
+    mLastStepPosition = mMovie->getCurrentTime();
+    CI_LOG_I( "POS AFTER STEP : " << mMovie->getCurrentTime() );
+    CI_LOG_I( "---------- NEW STEP FWD END ----------" );
+}
+
+void GstPlayerTestApp::update()
+{
+
+    testCurrentCase();
+
+    if( mMovie ) {
+        mFrameTexture = mMovie->getTexture();
+    }
+
+    static bool sPrintedDone = false;
+    if( ! sPrintedDone && mMovie->isDone() ) {
+        console() << "Done Playing" << std::endl;
+        sPrintedDone = true;
+    }
+}
+
+void GstPlayerTestApp::draw()
+{
+    gl::clear( Color( 0, 0, 0 ) );
+
+    if( mFrameTexture ) {
+        Rectf centeredRect = Rectf( mFrameTexture->getBounds() ).getCenteredFit( getWindowBounds(), true );
+        gl::draw( mFrameTexture, centeredRect );
+    }
+}
+
+void GstPlayerTestApp::keyDown( KeyEvent event )
+{
+    if( event.getCode() == KeyEvent::KEY_o ) {
+        fs::path moviePath = getOpenFilePath();
+        if( ! moviePath.empty() )
+            loadMovieFile( moviePath );
+    }
+    else if( event.getCode() == KeyEvent::KEY_p ) {
+        if( mMovie ) mMovie->stop();
+    }
+    else if( event.getCode() == KeyEvent::KEY_l ) {
+        if( mMovie ) mMovie->play();
+    }
+    else if( event.getCode() == KeyEvent::KEY_t ) {
+        if( mMovie ) mMovie->stepForward();
+    }
+    else if( event.getCode() == KeyEvent::KEY_f ) {
+        if( mMovie ) mMovie->stop();
+        mMovie->seekToFrame( sFrameNum++ );
+    }
+}
+
+void GstPlayerTestApp::loadMovieFile( const fs::path &moviePath )
+{
+    try {
+        // load up the movie, set it to loop, and begin playing
+        mMovie = qtime::MovieGl::create( moviePath );
+        mMovie->setLoop();
+        mMovie->play();
+	console() << "Playing: " << mMovie->isPlaying() << std::endl;
+    }
+    catch( ci::Exception &exc ) {
+        console() << "Exception caught trying to load the movie from path: " << moviePath << ", what: " << exc.what() << std::endl;
```

Comment:
Check out [`CI_LOG_EXCEPTION()`](https://github.com/cinder/Cinder/blob/master/include/cinder/Log.h#L375-L379), really handy for things like this.